### PR TITLE
Improve Typing

### PR DESCRIPTION
**Pull request by gronke**
Adds types to the following files:
- libiocage/lib/errors.py
- libiocage/lib/events.py

### CODE REVIEW

File: libiocage/lib/errors.py [hunk 1]
```diff
@@ -22,17 +22,27 @@
 # IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
 
+# MyPy
+import libzfs  # noqa: F401
```
Comment (igalic) [inline libiocage/lib/errors.py]: huh?

File: libiocage/lib/errors.py [hunk 2]
```diff
@@ -22,17 +22,27 @@
...
 
 class IocageException(Exception):
 
-    def __init__(self, message, errors=None, logger=None, level="error",
-                 append_warning=False, warning=None):
+    def __init__(
+        self,
+        message: str,
+        logger: 'libiocage.lib.Logger.Logger'=None,
+        level: str="error",
```
Comment (igalic) [inline libiocage/lib/errors.py]: this shouldn't be `str`, imo

File: libiocage/lib/errors.py [hunk 3]
```diff
@@ -220,35 +267,59 @@ def __init__(self, distribution_name, *args, **kwargs):
 
 class UnmountFailed(IocageException):
 
-    def __init__(self, mountpoint, *args, **kwargs):
+    def __init__(
+        self,
+        mountpoint: str,
```
Comment (igalic) [inline libiocage/lib/errors.py]: we should have a type for Unix paths

File: libiocage/lib/errors.py [hunk 4]
```diff
@@ -220,35 +267,59 @@ def __init__(self, distribution_name, *args, **kwargs):
...
         msg = f"Failed to unmount {mountpoint}"
         super().__init__(msg, *args, **kwargs)
 
 
 class MountFailed(IocageException):
 
-    def __init__(self, mountpoint, *args, **kwargs):
+    def __init__(
+        self,
+        mountpoint: str,
```
Comment (igalic) [inline libiocage/lib/errors.py]: a path type would be useful

File: libiocage/lib/events.py [hunk 5]
```diff
@@ -70,15 +74,13 @@ def __init__(self, message=None, **kwargs):
...
-                         error="failed",
-                         skipped="skipped",
-                         done="done",
-                         pending="pending"):
+    def get_state_string(
+        self,
+        error: str="failed",
+        skipped: str="skipped",
+        done: str="done",
+        pending: str="pending"
```
Comment (igalic) [inline libiocage/lib/events.py]: once again, i don't like that these are `str`

File: libiocage/lib/events.py [hunk 6]
```diff
@@ -92,7 +94,7 @@ def get_state_string(self,
         return pending
 
     @property
-    def type(self):
+    def type(self) -> str:
```
Comment (igalic) [inline libiocage/lib/events.py]: why._.

File: libiocage/lib/events.py [hunk 7]
```diff
@@ -115,51 +117,51 @@ def pending(self, state):
...
         if new_state is False:
-            self._stopped_at = timer()
+            self._stopped_at = float(timer())
 
         self._pending = new_state
         IocageEvent.PENDING_COUNT += 1 if (state is True) else -1
 
     @property
-    def duration(self):
+    def duration(self) -> typing.Optional[float]:
```
Comment (igalic) [inline libiocage/lib/events.py]: why is this float? doesn't time come in better types?

File: libiocage/lib/Types.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    unix_path = re.compile(r"/([^/\0]+/*)+")
+
+    def __init__(
+            self,
+            sequence: str
+    ) -> None:
+        if self.unix_path.fullmatch(sequence) is None:
+            raise TypeError("Invalid value for AbsolutePath")
+        super.__init__(sequence)
```
Comment (igalic) [inline libiocage/lib/Types.py]:
```
libiocage/lib/Types.py:40: error: No overload variant of "__init__" of "super" matches argument types [builtins.str]
```
ugh

File: libiocage/lib/errors.py (hunk 1 above)
Comment (gronke) [reply libiocage/lib/errors.py]: mypy complained about missing libzfs, so I've added it to the imports. Flake8 would now complain about the unused import unless we ignore F401.

File: libiocage/lib/Types.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import re
+from collections import UserString
+
+
+class AbsolutePath(UserString):
```
Comment (igalic) [inline libiocage/lib/Types.py]: I'm not sure `UserString` works here… i feel like the docs are lying to me… https://docs.python.org/3/library/collections.html#collections.UserString

File: libiocage/lib/events.py (hunk 7 above)
Comment (gronke) [reply libiocage/lib/events.py]: this is the delta of two timestamps. what other than float do you suggest to use?

File: libiocage/lib/events.py (hunk 6 above)
Comment (gronke) [reply libiocage/lib/events.py]: It's printed later. And it will be the reference for translations.

**Pull request merged**